### PR TITLE
RF copy_file() with a cleaner (cached) Repo method access

### LINKED ISSUE #6504: Copy-file is slow
I am in the process of updating HCP subsets that were ``copy-filed`` from the large HCP dataset, such as  https://github.com/datalad-datasets/hcp-structural-preprocessed. Here is one exemplary run commit used to find the relevant files and copyfile them: https://github.com/datalad-datasets/hcp-structural-preprocessed/commit/cfdced1d00300c3637f7de2396f42cf9cc699a81

Rerunning these run commands, and thus ``copy-file``, to update files and availability information works nicely. However, it is slow. On juseless, it takes more than a second to process one file, and some datasets require updates 50k or 70k files. I found https://github.com/datalad/datalad/pull/5599, which looks like it could substantially speed things up with improved caching but was closed mainly due to a lack of interest.

### PR DESCRIPTION
**Pull request by adswa**
This reincarnates #5599 in new hopes to fix #6504.

@mih's original description was:

> This is a first example of an approximation of an idea that I sketched in https://github.com/datalad/datalad/issues/5598
Instead of using a complicated beast made of dicts and more dicts, I RF'ed all relevant Repo method access into an Interface helper that uses lru_cache and other means to minimize the actual interaction with the underlying Repo instance.
It does that with awareness of what type of access is being made and only works within the context of copy_file().
I mostly did this to see if these task-focused/tailored API classes can make sense, and de-convolute primary command logic from performance aspects.
Happy to get feedback on the approach.

### CODE REVIEW

File: datalad/local/copy_file.py [hunk 1]
```diff
@@ -42,10 +42,151 @@
...
+        # TODO this could also be the place to stop lingering batch processes
+        if not self._tmpdir:
+            return
+
+        try:
+            self._tmpdir.rmdir()
+        except OSError as e:
+            lgr.warning(
+                'Failed to clean up temporary directory: %s',
+                exc_str(e))
```
Comment (mih) [inline datalad/local/copy_file.py:137]: This still needs to be converted to use CapturedException

Comment (adswa) [reply datalad/local/copy_file.py]: thanks for catching this! I pushed a fix

**Comment by adswa** [comment]
So far, the pattern of failures is the known one of #6492 plus a few github connection errors:

```
 ERROR: Could not install packages due to an OSError: HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /datalad/datalad-deprecated/archive/master.zip (Caused by ConnectTimeoutError(<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f7861d1b150>, 'Connection to github.com timed out. (connect timeout=15)'))
```

**Review by mih — APPROVED** [review]
LGTM! Thx!

**Pull request merged**